Work on `FlatsMatroid`

gmou3 opened the pull request:
Add class-optimized `_closure` and `_is_closed` methods.

Add input handling of a list of flats, and of a lattice of flats. Note that previously one could only define a `FlatsMatroid` from a dictionary of flats (indexed by their rank). The definition from a raw list requires more time upon definition as it computes and stores the lattice of flats from the input list. On the positive side, given this lattice, some methods become blazingly fast (e.g., `is_valid`, `whitney_numbers`).

The lattice of flats is now cached upon computation.

### :memo: Checklist

- [x] The title is concise and informative.
- [x] The description explains in detail what this PR is about.
- [ ] I have linked a relevant issue or discussion.
- [x] I have created tests covering the changes.
- [x] I have updated the documentation and checked the documentation preview.

gmou3 committed:
Work on `FlatsMatroid`

Add `_closure` and `_is_closed` methods.
Create subclass `LatticeOfFlatsMatroid` which results from an input of a
_list_ of flats (instead of a dict). This subclass requires more compute
upon definition as it computes and stores the lattice of flats from a
raw list. On the positive side, given this lattice, some methods become
blazingly fast (e.g., `is_valid`, `whitney_numbers`).

tscrim commented on src/sage/matroids/flats_matroid.pyx:
```diff
@@ -521,13 +562,243 @@ cdef class FlatsMatroid(Matroid):
+
+cdef class LatticeOfFlatsMatroid(FlatsMatroid):
+    r"""
+    INPUT:
```
```suggestion
cdef class LatticeOfFlatsMatroid(FlatsMatroid):
    r"""
    Matroid defined by its lattice of flats.
    
    INPUT:
```

tscrim commented on src/sage/matroids/flats_matroid.pxd:
```diff
@@ -7,7 +7,8 @@ cdef class FlatsMatroid(Matroid):
     cpdef full_rank(self)
-    cpdef _is_independent(self, F)
+    cpdef _closure(self, X)
+    cpdef _is_closed(self, X)
```
```suggestion
    cpdef bint _is_closed(self, X)
```
Perhaps it would be good to start adding such return type indicators in some key places? We probably have to see if the C code simplifies in some places.

gmou3 replied:
I am planning this on a massive scale on a future PR.

tscrim reviewed (changes requested):
I am not sure about having a full subclass. I think `FlatsMatroid` should be able to take in the lattice as input and store it (and/or have an optional argument to compute it). If the lattice has been computed (which the user could ask for at some point), then the lattice should be stored and the matroid should chose the code path that take advantage of that. With the proposed change, the flats matroid cannot take advantage of knowing the lattice structure.

It's a bit of an overhaul, but I don't think it will be too difficult. What do you think?

Otherwise, I have some minor comments.

gmou3 committed:
Remove LatticeOfFlatsMatroid & adapt FlatsMatroid

Move functionality of caching and using the lattice of flats (_L) to the
main FlatsMatroid class. Suggestion by tscrim.

tscrim reviewed (changes requested):
Thank you.

The only other thing is I would explicitly set `self._L = None` and check `self._L is None`. This is because (in principle) the lattice might be empty, which is planned to eventually be changed so that `bool(empty_lattice)` is `False`. This might cause an issue with the corner case of the empty matroid. Plus it makes it a bit more explicit that we are checking if something is initialized.

gmou3 committed:
Explicit check for lattice of flats

gmou3 committed:
Minor docstring improvements in matroid contructor

tscrim reviewed (approved):
I am happy with this as it stands. Positive review.
